src/runtime-rs/crates/resource/src/manager_inner.rs

```diff
@@ -219,10 +240,54 @@ impl ResourceManagerInner {
         spec: &oci::Spec,
     ) -> Result<Vec<Arc<dyn Volume>>> {
         self.volume_resource
-            .handler_volumes(&self.share_fs, cid, spec)
+            .handler_volumes(
+                &self.share_fs,
+                cid,
+                spec,
+                self.device_manager.clone(),
+                &self.sid,
+            )
             .await
     }
 
+    pub async fn handler_devices(
+        &self,
+        _cid: &str,
+        linux: &Linux,
+        devices_agent: &mut Vec<Device>,
+    ) -> Result<()> {
+        for d in linux.devices.iter() {
+            // generate device info from oci spec
+            let mut device_info = self
+                .device_manager
+                .read()
+                .await
+                .new_device_info_from_oci(d, None)?;
+            let dev_class = match device_info.dev_type.as_str() {
+                "b" => Block,
+                _ => Undefined,
+            };
+            // attach device
+            let device_id = self
+                .device_manager
+                .write()
+                .await
+                .try_add_device(&mut device_info, &dev_class)
+                .await
+                .context("failed to add device")?;
+            // generate agent device info
+            let device = self
+                .device_manager
+                .read()
+                .await
+                .generate_agent_device(device_id, &dev_class)
+                .await
+                .context("failed to generate agent device")?;
+            devices_agent.push(device);
```

Comment:
I think the line from 261-285 should wrap in the device manager.
DeviceManager function:
```
impl DeviceManager {
    pub fn new_agent_device(device: &oci::LinuxDevice,) -> Result<AgentDevice> {
        ...
    }
} 
```
Call from handler_devices
```
for d in linux.devices.iter() {
    let device = self.device_manager.write.await.new_agent_device(d).context("")?;
    device_agent.push(device);   
}
```

Reply:
Yeah, wrap it into DeviceManager is OK.
And it's better to change the generate_agent_device(device_id, &dev_class) to generate_agent_device(d). 
If so, no need to add a new method.